src/terminal.rs

```diff
@@ -0,0 +1,381 @@
+    new_total_width: usize,
+    new_total_width_to_cursor: usize,
+    new_num_lines: usize,
+}
+
+impl Metrics {
+    pub fn new(
+        prompt: String,
+        buf: &Buffer,
+        cursor: &Cursor,
+        autosuggestion: Option<&Buffer>,
+    ) -> io::Result<Self> {
+        let width = util::terminal_width()?;
+        let prompt_width = util::last_prompt_line_width(&prompt);
+        let buf_width = buf.width();
+
+        let buf_widths = match autosuggestion {
+            Some(suggestion) => suggestion.width(),
+            None => buf_width,
+        };
+
+        // Width of the current buffer lines (including autosuggestion) from the start to the cursor
+        let buf_widths_to_cursor = match autosuggestion {
+            // Cursor might overrun autosuggestion with history search.
+            Some(suggestion) if cursor.char_vec_pos() < suggestion.num_chars() => {
+                suggestion.range_width(0, cursor.char_vec_pos())
+            }
+            _ => buf.range_width(0, cursor.char_vec_pos()),
+        };
+        // Total number of terminal spaces taken up by prompt and buffer
+        let new_total_width = Metrics::calc_width(prompt_width, &buf_widths, width);
+        let new_total_width_to_cursor =
+            Metrics::calc_width(prompt_width, &buf_widths_to_cursor, width);
+
+        let new_num_lines = (new_total_width + width) / width;
+
+        Ok(Metrics {
+            width,
+            prompt_width,
+            new_total_width,
+            new_total_width_to_cursor,
+            new_num_lines,
+        })
+    }
+
+    pub fn max_x_dimensions(&self) -> bool {
+        self.new_total_width % self.width == 0
+    }
+
+    /// Move the term cursor to the same line as the prompt.
+    fn calc_width(prompt_width: usize, buf_widths: &[usize], terminal_width: usize) -> usize {
+        let mut total = 0;
+        for line in buf_widths {
+            if total % terminal_width != 0 {
+                total = ((total / terminal_width) + 1) * terminal_width;
+            }
+            total += prompt_width + line;
+        }
+        total
+    }
+}
+
+/// An interface to simplify writing to and clearing the terminal. This is where most
+/// of the display logic for interfacing with the console library exists.
+pub struct Terminal<'a> {
+    out: &'a mut dyn io::Write,
+    // The line of the cursor relative to the prompt. 1-indexed.
+    // So if the cursor is on the same line as the prompt, `term_cursor_line == 1`.
+    // If the cursor is on the line below the prompt, `term_cursor_line == 2`.
+    term_cursor_line: usize,
+    // Last string that was colorized and last colorized version.
+    color_lines: Option<(String, String)>,
+    // A closure that is evaluated just before we write to out.
+    // This allows us to do custom syntax highlighting and other fun stuff.
+    closure: Option<ColorClosure>,
+    // Use the closure if it is set.
+    use_closure: bool,
+    buf: &'a mut String,
```

Comment:
Should this be a &'a mut str?

Reply:
I tried it out, because this is the underlying terminal buffer and methods like write! are called with it I think it has to be String.